### ISSUE TITLE
Should resource.NewIndex return a concrete type?

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#109]
@rs, I am raising this as a potential enchantment only.

It appears to be an idiom in Go to return concrete types, and accept interfaces. `resource.NewIndex()` is breaking this idiom. I see the following benefits for correcting it:
- `index.Compile()` would no longer be a "hidden" function, so user would not have to type-cast the returned index to a Compiler instance if using the returned Index without the rest handlers.
- The documentation in editors (i.e. with go-to-definitions and auto-complete) would be able to display help-text when calling methods on the returned index.

These are not huge benefits, but if there is no obvious downside to the change, except for a limited change to the public API, it might be worth-while. There are also some consequences that could come out of _how_ we fix this that we need to consider.

- `resource.Resource` implements the Index interface. Could/should Bind accept a version of the Index interface?
- Where the Index interface is used, du we require all of the functions, or generally just the Getter functions? Should the interface be split?

# Background: General information on the idiom
Some blogs on the subject:
- http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/
- https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8

TL;DR with some added personal experience:

Generally, returning the concrete type makes sense because:
- When we write a function, we always know what type it will return (when we don't, such as for errors, returning an interface still make sense).
- When we return the concrete type, users can access all public functions on the returned type.
- Documentation in auto-complete and go-to-definition is generally better.

Likewise, it makes sense to _accept_ interfaces because:
- We allow for a wider usage of the API, as users may provide their own implementations, or wrapped implementations.
- We achieve better documentation, since the function declaration itself can _document_ which methods we are interested in using on the passed in type.
- We allow for mocking during testing.

When breaking the idiom, it should be because some of the points above don't match with our goal. E.g. accepting an interface for `MaxInt(a, b int) int` is clearly bananas.

**Comment by rs** [comment]
+1

**Comment by smyrman** [comment]
I have been thinking a little bit about this for a while, and I have come to a perhaps slightly unexpected conclusion.

I think that an uncompiled Index and a Compiled index is not really the same thing, and does not have the same interface. I will collaborate a little-bit on this...

An uncompiled Index has the following use-case and properties:
- You want to be able to bind new resources against it.
- You DON'T want to be fetching any actual resources _items_ from it.

- You want to eventually compile it.

A compiled Index has the following use-cases and properties:
- You want to be

**Comment by smyrman** [comment]
PS! This design also completely _guards_ against users using an _uncompiled_ index by mistake, as doing so would now be a _compiler error_.

**Comment by rs** [comment]
I like the idea.
